### DISCUSSION TITLE
OSIV 옵션에 대해서 설명해주세요.

### DISCUSSION THREAD

**Discussion by nak-honest** [woowacourse-6th-computer-sign-pen/comsa-interview-study#10] [JPA] [answered]
OSIV 옵션에 대해서 설명해주세요.

**Comment by slimsha2dy** [comment, marked as answer]
**답변자: 낙낙(이낙헌)**

질문자: OSIV 옵션에 대해 설명해 보시겠어요?

답변자: OSIV는 'Open Session In View'의 약자로, 말 그대로 뷰(View)에서도 세션(Session) 정보를 볼 수 있다는 의미입니다. 일반적으로 OSIV가 비활성화된 경우를 먼저 설명드리면, 서비스 메소드가 호출될 때 트랜잭션이 시작되고 이때 영속성 컨텍스트가 열립니다. 비즈니스 로직을 수행하고 서비스 메소드가 끝날 때, 즉 반환할 때 영속성 컨텍스트가 닫히며 동시에 DB 커넥션과 트랜잭션도 닫히는 것이 일반적인 동작입니다. 하지만 OSIV를 true로 설정하면 서비스 메소드 바깥에서도 영속성 컨텍스트가 살아 있게 됩니다.

영속성 컨텍스트는 스프링 인터셉터를 이용해 시작되고, 생명 주기가 길어집니다. 그래서 뷰나 컨트롤러에서도 영속성 컨텍스트를 이용해 추가적인 읽기 작업이 가능합니다. OSIV는 이를 위해 만들어진 것이죠. 예를 들어, Lazy Loading을 설정한 상태에서 서비스에서 지연 로딩된 엔티티를 조회하지 않은 채 반환했을 때, OSIV가 켜져 있으면 서비스 메소드 밖에서도 컨트롤러나 뷰에서 지연 로딩된 데이터를 읽을 수 있습니다. 기본적으로 스프링에서는 OSIV 옵션이 true로 설정되어 있습니다.

질문자: 그러면 컨트롤러에서 읽기 위한 목적인가요?

답변자: 컨트롤러뿐만 아니라, 앞서 말씀드린 것처럼 뷰에서도 읽을 수 있도록 지원해 줍니다. 다만 뷰 쪽은 제가 REST API로만 구현해 왔기 때문에, 정확한 동작 원리는 자세히 알지 못합니다.

질문자: 단점은 뭐가 있을까요?

답변자: 단점은 DB 커넥션을 오래 유지하게 되어 커넥션 풀 고갈 문제가 발생할 수 있다는 것입니다. 서비스뿐만 아니라 더 길게 커넥션을 잡기 때문에 이러한 문제가 생길 수 있습니다.

질문자: 시간이 많이 길어지나요?

답변자: 정확한 수치는 모르겠지만, 그렇게 길 것 같지는 않습니다.

질문자: 수치라기보다는 영향이 갈 정도로 길어질 수 있나요? 컨트롤러 로직만 있으면 별로 안 길 것 같은데요.

답변자: 저도 크게 크리티컬한 문제는 아닐 것 같아요. 결국 네트워크로 나가기 전에 영속성 컨텍스트를 닫을 테니까요. 하지만 뷰에서는 디스크 I/O 작업이 들어가기 때문에, 그 작업 동안에도 DB 커넥션을 유지하면 성능 저하가 발생할 수 있다고 생각합니다. 그러나 REST API로 구현되어 있을 때는 해당되지 않으니 큰 문제는 없을 것 같습니다.

질문자: REST API에서는 그런 일이 없을까요?

답변자: REST API에서는 동작 원리가 어디서 끊기는지에 따라 다를 것 같습니다.

질문자: 영속성 컨텍스트가 인터셉터에서 생성되고 소멸된다고 생각하면 어떤 문제가 있을까요?

답변자: 인터셉터에서 종료될 때까지 영속성 컨텍스트가 유지된다면, 성능적인 문제는 크지 않을 것 같습니다. 하지만 제가 경험한 바로는, DB를 리더와 라이터로 분리했을 때 트랜잭션의 readOnly 옵션에 따라 데이터 소스를 결정하는 로직이 있었습니다. OSIV가 true로 설정된 상태에서 인터셉터를 통해 리포지토리를 직접 호출하면, 서비스 메소드가 호출되기 전에 트랜잭션이 생성되고 유지됩니다. 이로 인해 원래는 쓰기 트랜잭션이어야 하는데, 이미 readOnly로 설정되어 읽기 전용 데이터베이스에 접근하는 문제가 발생했습니다. 이런 경우에는 주의가 필요합니다.

성능상의 문제점은 깊게 파보지 않아서, 혹시 자세히 설명해 주실 수 있을까요?

질문자: 깊게 파보진 않았지만, 일반적으로 성능을 고려할 때 커넥션을 유지하는 시간을 최소화하려고 외부 API 호출을 트랜잭션 밖으로 빼잖아요. 그런데 OSIV가 켜져 있으면 트랜잭션을 밖으로 빼도 큰 의미가 없을 수 있어요.

답변자: 맞는 말씀입니다. 예를 들어 AOP 처리를 해당 단에서 한다거나 하면 그럴 수 있겠네요.

질문자: 지금 저희 구조처럼 외부에서 S3에 접근하는 걸 트랜잭션 밖으로 빼더라도, OSIV가 켜져 있으면 문제가 생길 수 있어요.

답변자: 그렇네요. 외부 API를 트랜잭션 밖에서 호출하지 않는다면 성능 저하는 크지 않을 것 같습니다.

질문자: 그렇죠. 외부 로직이 복잡하지 않다면 큰 문제는 없을 것 같아요.

답변자: REST API 말고 뷰를 직접 렌더링하는 경우에는 어떨까요?

질문자: 찾아보니 뷰 렌더링이 API보다 훨씬 오래 걸린다고 하더라고요.

답변자: 그렇다면 디스크 I/O 작업 때문일 것 같습니다. 그런 이유로 성능 저하가 있을 수 있겠네요. OSIV 관련해서 더 생각나는 질문 있으신가요?

질문자: 제가 찾아보니 OSIV가 안티 패턴이라는 의견도 많더라고요.

답변자: 안티 패턴이라니, 커넥션 고갈이나 성능 문제 외에 다른 이유도 있나요?

질문자: 스프링 부트 이슈로 봤는데, 내용이 많아서 다 보진 못했어요. 초보 개발자가 JPA를 사용할 때 OSIV가 켜져 있으면 지연 로딩에 대해 인지하지 못할 수 있다는 내용이 있더라고요. 또 "JPA는 초보 개발자만 사용하냐"는 반박도 있었고요. 관심 있으시면 한 번 읽어보세요.

답변자: 저도 궁금한 게 있습니다. OSIV를 사용하면 엔티티가 컨트롤러까지 넘어가서, 컨트롤러에서 엔티티를 조작하거나 지연 로딩을 발생시킬 수 있습니다. 그런데 서비스의 역할이 컨트롤러로부터 도메인을 캡슐화하는 것이기 때문에, 보통 DTO로 감싸서 전달하잖아요. 그래서 OSIV를 사용해 컨트롤러에서 도메인을 직접 사용하는 것은 안티 패턴으로 볼 수 있지 않을까요? 도메인의 변화가 컨트롤러까지 전달되니까요.

질문자: 저도 정확히는 모르지만, 뷰를 사용할 때 도메인을 직접 조작하지 않고 그냥 뷰에 넘길 것 같아요.

답변자: DTO를 뷰에 넘길 수도 있지 않나요?

질문자: 그렇죠. 하지만 OSIV를 사용해서 뷰에서 지연 로딩을 처리한다면, 컨트롤러에서 도메인의 내부 구현을 알아야 하고, 그걸 조작해서 뷰로 넘기는 것보다는 그냥 도메인을 뷰로 넘기지 않을까 싶어요. 추측이긴 하지만요.

답변자: 도메인을 뷰에 그대로 넘기는 것도 계층 간의 경계를 넘는 것이니 문제일 것 같습니다.

질문자: 맞아요. 좋은 방법은 아닌 것 같아요.

답변자: DTO로 넘기는 게 더 좋아 보입니다. 뷰까지 도메인을 넘기는 것은 예전에 저도 해봤는데, 리뷰에서 지적받은 적이 있어요. 당시에는 콘솔 프로그램이라 넘어갔지만, 웹 프로그램에서는 위험할 수 있다고 생각합니다. 아무튼 JPA를 잘 모르는 경우에는 OSIV를 false로 설정하는게 오히려 더 안전할 것 같네요. 어떤 문제가 발생할지 모르니까요.
